src/toxcore/dht/server.rs

```diff
@@ -205,29 +217,29 @@ impl Server {
                 match payload {
                     Ok(DhtRequestPayload::NatPingRequest(pl)) => {
                         debug!("Received nat ping request");
-                        let client = self.create_client(&addr, dr.spk);
-                        self.handle_nat_ping_req(client, dr, pl)
+                        self.handle_nat_ping_req(dr, pl, addr)
                     },
                     Ok(DhtRequestPayload::NatPingResponse(pl)) => {
                         debug!("Received nat ping response");
-                        if let Some(client) = self.get_client(&dr.spk) {
-                            self.handle_nat_ping_resp(client, dr, pl)
+                        let state = self.state.read();
+                        if let Some(client) = self.get_client(&dr.spk, &state.peers_cache) {
```

Comment:
You must call `handle_*(packet, addr)` here, while you take locks inside handle_* function.